src/expr/src/relation/mod.rs

```diff
@@ -208,6 +208,18 @@ pub enum MirRelationExpr {
         /// Columns to arrange `input` by, in order of decreasing primacy
         keys: Vec<Vec<MirScalarExpr>>,
     },
+    /// Declares that `key` is a primary key for `input`.
+    /// Should be used *very* sparingly, and only if there's no plausible
+    /// way to derive the key information from the underlying expression.
+    /// The result of declaring a key that isn't actually a key for the underlying expression is undefined.
+    ///
+    /// There is no operator rendered for this IR node; thus, its runtime memory footprint is zero.
+    DeclareKey {
+        /// The source collection
+        input: Box<MirRelationExpr>,
+        /// The set of columns in the source collection that form a key.
+        key: Vec<usize>,
```

Comment:
This could be `DeclareKeys` with `keys: Vec<Vec<usize>>`, which would support e.g. various `UNIQUE` constraints for inputs. General the inferred "key" structure produces potentially multiple keys. Can certainly add at a later date, but wanted to point it out!

Reply:
Right, but you could stack multiple `DeclareKey` in that case. Not sure whether one approach is any better than the other. I suppose if you have a bunch of keys, `foo.declare_keys(vec![a, b, c])` is easier to write than `foo.declare_key(a).declare_key(b).declare_key(c)`, so sure, I can implement it.